src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -396,12 +476,66 @@ hardLinkRunFiles ::
 hardLinkRunFiles reg hfs hbio dur sourceRunFsPaths targetRunFsPaths = do
     let sourcePaths = pathsForRunFiles sourceRunFsPaths
         targetPaths = pathsForRunFiles targetRunFsPaths
-    sequenceA_ (hardLinkTemp <$> sourcePaths <*> targetPaths)
-    hardLinkTemp (runChecksumsPath sourceRunFsPaths) (runChecksumsPath targetRunFsPaths)
-  where
-    hardLinkTemp sourcePath targetPath = do
-        allocateTemp reg
-          (FS.createHardLink hbio sourcePath targetPath)
-          (\_ -> FS.removeFile hfs targetPath)
-        when (dur == HardLinkDurable) $
-          FS.synchroniseFile hfs hbio targetPath
+    sequenceA_ (hardLinkTemp reg hfs hbio dur <$> sourcePaths <*> targetPaths)
+    hardLinkTemp reg hfs hbio dur (runChecksumsPath sourceRunFsPaths) (runChecksumsPath targetRunFsPaths)
+
+{-------------------------------------------------------------------------------
+  Hard link file
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE
+  hardLinkTemp ::
+      TempRegistry IO
+    -> HasFS IO h
+    -> HasBlockIO IO h
+    -> HardLinkDurable
+    -> FS.FsPath
+    -> FS.FsPath
+    -> IO ()
+  #-}
+-- | @'hardLinkTemp' reg hfs hbio dur sourcePath targetPath@ creates a hard link
+-- from @sourcePath@ to @targetPath@.
+hardLinkTemp ::
```

Comment:
Let's call it just `hardLink` instead. The `Temp` prefix is going to be out of date soon.